src/utils/formModeUtils.ts

```diff
@@ -0,0 +1,158 @@
+  const isStatusViewOnlyForUser = ["Submitted", "In Review", "Approved"].includes(formStatus);
+
+  if (formStatus !== "New" && !formBelongsToUser) {
+    return FormModes.UNAUTHORIZED;
+  }
+  if (isStatusViewOnlyForUser) {
+    return FormModes.VIEW_ONLY;
+  }
+  if (EditStatuses.includes(formStatus)) {
+    return FormModes.EDIT;
+  }
+  return FormModes.UNAUTHORIZED;
+};
+
+/**
+ * Calculate the form mode for a Federal Lead
+ * NOTE:
+ *  - This is a private helper function for getFormMode
+ *
+ * @param {User} user - The current user
+ * @param {Application} data - The current application/submission
+ * @returns {FormMode} - Form mode corresponding to the given form status for a Federal Lead.
+ */
+const _getFormModeForFederalLead = (
+  user: User,
+  data: Application
+): FormMode => {
+  if (user?.role !== "FederalLead") {
+    return FormModes.UNAUTHORIZED;
+  }
+
+  const { status: formStatus } = data || {};
+  const isStatusViewOnlyForLead = ["In Progress", "Approved", "Rejected"].includes(formStatus);
+
+  if (ReviewStatuses.includes(formStatus)) {
+    return FormModes.REVIEW;
+  }
+  if (isStatusViewOnlyForLead) {
+    return FormModes.VIEW_ONLY;
+  }
+
+  return FormModes.UNAUTHORIZED;
+};
+
+/**
+ * Calculate the form mode for an Organization Owner
+ * NOTE:
+ *  - This is a private helper function for getFormMode
+ *
+ * @param {User} user - The current user
+ * @param {Application} data - The current application/submission
+ * @returns {FormMode} - Form mode corresponding to the given form status and organization owner.
+ */
+const _getFormModeForOrgOwner = (
+  user: User,
+  data: Application
+): FormMode => {
+  const belongsToSameOrg = user.organization?.orgID === data?.organization?._id;
+  const orgRole = belongsToSameOrg ? user.organization?.orgRole : null;
+  if (orgRole !== "Owner") {
+    return FormModes.UNAUTHORIZED;
+  }
+
+  const { status: formStatus } = data || {};
+  const formBelongsToUser = data?.applicant?.applicantID === user?.["_id"];
+  const isStatusViewOnlyForOrgOwner = ["Submitted", "In Review", "Approved"].includes(formStatus);
+
+  if (!formBelongsToUser || isStatusViewOnlyForOrgOwner) {
+    return FormModes.VIEW_ONLY;
+  }
+  if (EditStatuses.includes(formStatus)) {
+    return FormModes.EDIT;
+  }
+  return FormModes.UNAUTHORIZED;
+};
+
+/**
+ * Get updated form mode based on role, organization role, and form status
+ * NOTE:
+ * - Depends on the following private helper functions:
+ *    _getFormModeForUser,
+ *    _getFormModeForFederalLead,
+ *    _getFormModeForOrgOwner
+ *
+ * @param {User} user - The current user
+ * @param {Application} data - The current application/submission
+ * @returns {FormMode} - Updated form mode based on role, organization role, and form status
+ */
+export const getFormMode = (
+  user: User,
+  data: Application,
+): FormMode => {
+  if (!user || (!user?.role && !user?.organization?.orgRole) || !data) {
```

Comment:
A user should always have a role. This scenario should never occur, but we may want to consider missing `user.role` as unauthorized, even if they have an organizational role.

Reply:
Updated to return unauthorized when user role does not exist. Thanks.